lib/utils/emitWarning.mjs

```diff
@@ -0,0 +1,73 @@
+import process from 'node:process';
+
+// Each code must be unique across all warning types.
+const CODES = {
+	COMMONJS_PLUGINS: '001',
+	COMMONJS_NODEJS_API: '002',
+	RESULT_OUTPUT_PROPERTY: '003',
+	GITHUB_FORMATTER: '004',
+	CONTEXT_FIX: '005',
+	RULE: '006', // either removal or renaming
+	REPORT_AMBIGUOUS_POSITION: '007',
+	SUPPRESSIONS: '008', // experimental
+};
+
+const emittedWarnings = new Set();
+
+/**
+ * Emit a warning with the specified type.
+ *
+ * @param {string} message
+ * @param {'DeprecationWarning' | 'ExperimentalWarning'} type
```

Comment:
[suggestion] How about adding a link to the type's reference in the Node.js official documentation?

```suggestion
 * @param {'DeprecationWarning' | 'ExperimentalWarning'} type - See https://nodejs.org/api/process.html#nodejs-warning-names
```